### ISSUE TITLE
Making `ImageSpec` a `value_type` fails `static_assert` about expected and actual size of generated Rust struct.

### ISSUE THREAD

**Issue by virtualritz** [anderslanglands/bbl-oiio#3]
`ImageSpec` has a ton of fields which are public and consequently no getters and setters (for most).

Accordingly (and also for ergonomics on the Rust side) I would like to make this a `value_type`.

Adding `.value_type()` after [line 32 in `imageio.cpp`](https://github.com/anderslanglands/bbl-oiio/blob/main/bind/imageio.cpp#L32) gets me:

```
error: failed to run custom build command for `oiio v0.1.0 (/home/moritz/code/crates/bbl-oiio)`

Caused by:
  process didn't exit successfully: `/home/moritz/code/crates/bbl-oiio/target/debug/build/oiio-7aeb46ff0e5158a1/build-script-build` (exit status: 101)
  --- stdout
  cargo::rustc-link-arg=-lstdc++
  running "cmake" "-G" "Ninja" "-DCMAKE_BUILD_TYPE=Release" "-S bbl-oiio" "-B /home/moritz/code/crates/bbl-oiio/target/debug/build/oiio-0aeb4a980c00a1b4/out/build" "-DCMAKE_INSTALL_PREFIX=/home/moritz/code/crates/bbl-oiio/target/debug/build/oiio-0aeb4a980c00a1b4/out" "-DBBL_LANGUAGES=rust"
  -- Configuring done (0.1s)
  -- Generating done (0.0s)
  -- Build files have been written to: /home/moritz/code/crates/bbl-oiio/target/debug/build/oiio-0aeb4a980c00a1b4/out/build
  running "cmake" "--build" "/home/moritz/code/crates/bbl-oiio/target/debug/build/oiio-0aeb4a980c00a1b4/out/build" "--target" "install"
  [1/4] Building CXX object CMakeFiles/oiio-c.dir/oiio-c.cpp.o
  FAILED: CMakeFiles/oiio-c.dir/oiio-c.cpp.o 
  /usr/bin/c++ -DNOMINMAX -I/home/moritz/code/crates/bbl-oiio/bbl-oiio -isystem /usr/include/opencv4 -isystem /usr/include/Imath -O3 -DNDEBUG -MD -MT CMakeFiles/oiio-c.dir/oiio-c.cpp.o -MF CMakeFiles/oiio-c.dir/oiio-c.cpp.o.d -o CMakeFiles/oiio-c.dir/oiio-c.cpp.o -c /home/moritz/code/crates/bbl-oiio/target/debug/build/oiio-0aeb4a980c00a1b4/out/build/oiio-c.cpp
  /home/moritz/code/crates/bbl-oiio/target/debug/build/oiio-0aeb4a980c00a1b4/out/build/oiio-c.cpp:209:55: error: static assertion failed: size of value type does not match
    209 | static_assert(sizeof(oiio_ImageSpec_t_bbl_size_check) == sizeof(OIIO::ImageSpec), "size of value type does not match");
        |               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
  /home/moritz/code/crates/bbl-oiio/target/debug/build/oiio-0aeb4a980c00a1b4/out/build/oiio-c.cpp:209:55: note: the comparison reduces to ‘(8 == 160)’
  ninja: build stopped: subcommand failed.

  --- stderr
  thread 'main' panicked at /home/moritz/.cargo/git/checkouts/bbl-build-rs-5843947e05bbdf87/fd2a2d1/src/lib.rs:267:9:
  command did not execute successfully, got: exit status: 1
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

For some reason `babble` seems to 'ignore' the request and still generate an opaque type instead but then  (correctly) check for a `value_type`.

`ImageSpec` members that are not PODs are `TypeDesc` and `ParamValueList` (`Vec<ParamValue>`). For all of these `babble` generates bindings just fine.

**Comment by anderslanglands** [comment]
Hmm I need to guard against this better and generate a better error message.

Basically, you can't bind ImageSpec as a value type because it contains non-trivial members. Trying to do so would make bad things™ happen, so it has to be opaque ptr.

It would be nice if there was an easy way to automate generating getters and setters for fields, but for now you'll have to do this manually, see https://github.com/anderslanglands/babble?tab=readme-ov-file#extension-functions

**Comment by virtualritz** [comment]
> Hmm I need to guard against this better and generate a better error message.
> 
> Basically, you can't bind ImageSpec as a value type because it contains non-trivial members. Trying to do so would make bad things™ happen, so it has to be opaque ptr.

What bad things would happen? Anything non-trivial would simply be a `void*` pointer.
Then one only needs to add getters and setters for non-trivial types.

I would really like this limitation lifted on the `babble` side and instead have this introduce limitations on the how-to-use-the-binding side:

* When creating such a type you must call a C++ constructor (or add one, yourself).
* Same when tearing it down.

I n return you get a struct where all trivial members can be accessed directly and w/o any need for boilerplate.

`ImageSpec` is a good example because the Rust version of my bindings will have two types. `ImageSpecConfig` which mirrors the C++ `struct`/`class` on the Rust side so everything can be accessed w/o any need for getter or setters there, just in C++.

```rust
pub struct ImageSpec {
    ptr: *mut oiio_ImageSpec_t,
}

pub struct ImageSpecConfig {
    /// Origin (upper left corner) of pixel data.
    x: i32,
    /// Origin (upper left corner) of pixel data.
    y: i32,
    /// Origin (upper left corner) of pixel data.
    z: i32,
    ...
}
```

What is returned or fed to the high level wrapper methods is an `ImageSpec` (which is opaque as you can see above).
However, methods that require an `ImageSpec` take `impl Into<ImageSpec>` instead of the actual type. A user can then either feed an opaque existing `ImageSpec` or an `ImageSpecConfig` which implements that trait.

Now implementing is just assignments; only for non-trivial members setters need to be called (and they actually exist already mostly, on the original OIIO `ImageSpec` class.

When `ImageSpec`s are returned they can be converted `Into<ImageSpecConfig>`s, accordingly, if neccessary.

`ImageSpec` will still have getters and setters too, but these can be avoided by calling `ImageSpec::into::<ImageSpecConfig>()` and then working with the members of that. I.e. incur a slight overhead for the copy of the struct for ergonomics -- or call getters and setters.

That way a user can select from the best of ergonomics, safety and speed.

I can certainly do all this right now but I have to write a ton of boilerplate (did you see how many trivial members `ImageSpec` has? 20. That's 40 getters/setters. Come on now! :smirk:)

Even autogenerating them -- why would you? If you want to keep the high level wrapper type opaque on the Rust side, you'd make the getters/setters there, not on the C++ side.

**Comment by anderslanglands** [comment]
> What bad things would happen? Anything non-trivial would simply be a void* pointer.

what does the void* point to?  Just to be clear, are you suggesting:
```
struct oiio_ImageSpec {
    int x;
    int y;
    // ...
    void* channelnames;
    void* extra_attribs;
};
```

**Comment by virtualritz** [comment]
> what does the void* point to? Just to be clear, are you suggesting:
> 
> ```
> struct oiio_ImageSpec {
>     int x;
>     int y;
>     // ...
>     void* channelnames;
>     void* extra_attribs;
> };
> ```

To nothing you can do anything with in Rust. Maybe `void*` is rather a bad idea and just `[u8]` of the resp. needed size/padding is better.

Basically it's *blind data* that is just there so the C++ and the Rust side of a struct size-match.
And of course, it may actually contain something meaningful but *only for C++*.
I.e. whatever is there can only be manipulated through some sort of getters/setters that exist on the C++ side.

E.g. if you get an `const* oiio_ImageSpec` pointer from somewhere and you dereference that, you would end up with something on the Rust side where you can savely modify all *trivial* fields *direcly* and all *non-trivial* ones *indirectly*. I.e. using getters/setters that existed or that you defined on the C++/`babble`-`bind` side.

I think `bindgen` uses something along those lines to pad some structs to be 100% equivalent to what they reflect, on the C side.

**Comment by virtualritz** [comment]
TLDR; no need for getters/setters for trivial members with this approach.
Neither manual-written ones nor `babble`-generated ones.

**Comment by anderslanglands** [comment]
So in previous babble versions (and in cppmm) we had something called opaque bytes which essentially did this, but for a whole type rather than individual fields. You can still see references to it in the code, and even declare a type to be opaque bytes I think it just does nothing. It's a todo to resurrect that (my first implementation in babble was causing stack corruption on Linux which I couldn't immediately figure out).

There's some fiddlyness around passing values back and forth to avoid destructors being run when you move a C++ value into the C bytes you're returning but I think we can work around that with unions.

A bigger problem is that this is a massive footgun since there's no move semantics in C. What's more there are types that are explicitly non-trivially moveable too (e.g. some implementations of std::string are self-referential in the SSO case). There's no good way of representing this in either C or Rust. 

So ultimately I'm still not convinced that the annoyance of having to do getters and setters is worth opening the UB can of worms that this implies.

**Comment by anderslanglands** [comment]
To be clear: I understand the desire, and even for types that are completely opaque it would be nice for performance reasons to pass them on the stack rather than by pointer, this just gets very hairy very quickly.
